### PR TITLE
Broadcast matrix inputs to Gemm

### LINKED ISSUE #984: Gemm fails with simple broadcasting case
```python
import aesara
import aesara.tensor as at
import numpy as np

b_v = np.ones((1, 100))
w_v = np.ones((6, 100))
x_v = np.ones((10, 6))

x = at.matrix("x")
b = at.matrix("b")
w = at.matrix("w")

out = at.dot(x, w) + b

# It works fine if mode="FAST_COMPILE", which avoids introducing the GEMM Op
f = aesara.function([x, b, w], out)  

aesara.dprint(f, print_type=True)
```
```
Gemm{no_inplace} [id A] <TensorType(float64, (None, None))> ''   0
 |b [id B] <TensorType(float64, (None, None))>
 |TensorConstant{1.0} [id C] <TensorType(float64, ())>
 |x [id D] <TensorType(float64, (None, None))>
 |w [id E] <TensorType(float64, (None, None))>
 |TensorConstant{1.0} [id C] <TensorType(float64, ())>
```

```python
f(x_v, b_v, w_v)
```
```python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File ~/miniconda3/envs/aesara/lib/python3.10/site-packages/aesara/compile/function/types.py:964, in Function.__call__(self, *args, **kwargs)
    962 try:
    963     outputs = (
--> 964         self.fn()
    965         if output_subset is None
    966         else self.fn(output_subset=output_subset)
    967     )
    968 except Exception:

ValueError: Shape mismatch: x has 10 rows but z has 1 rows

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
Input In [6], in <cell line: 19>()
     15 f = aesara.function([x, b, w], out)
     17 aesara.dprint(f, print_type=True)
---> 19 f(x_v, b_v, w_v)

File ~/miniconda3/envs/aesara/lib/python3.10/site-packages/aesara/compile/function/types.py:977, in Function.__call__(self, *args, **kwargs)
    975     if hasattr(self.fn, "thunks"):
    976         thunk = self.fn.thunks[self.fn.position_of_error]
--> 977     raise_with_op(
    978         self.maker.fgraph,
    979         node=self.fn.nodes[self.fn.position_of_error],
    980         thunk=thunk,
    981         storage_map=getattr(self.fn, "storage_map", None),
    982     )
    983 else:
    984     # old-style linkers raise their own exceptions
    985     raise

File ~/miniconda3/envs/aesara/lib/python3.10/site-packages/aesara/link/utils.py:538, in raise_with_op(fgraph, node, thunk, exc_info, storage_map)
    533     warnings.warn(
    534         f"{exc_type} error does not allow us to add an extra error message"
    535     )
    536     # Some exception need extra parameter in inputs. So forget the
    537     # extra long error message in that case.
--> 538 raise exc_value.with_traceback(exc_trace)

File ~/miniconda3/envs/aesara/lib/python3.10/site-packages/aesara/compile/function/types.py:964, in Function.__call__(self, *args, **kwargs)
    961 t0_fn = time.time()
    962 try:
    963     outputs = (
--> 964         self.fn()
    965         if output_subset is None
    966         else self.fn(output_subset=output_subset)
    967     )
    968 except Exception:
    969     restore_defaults()

ValueError: Shape mismatch: x has 10 rows but z has 1 rows
Apply node that caused the error: Gemm{no_inplace}(b, TensorConstant{1.0}, x, w, TensorConstant{1.0})
Toposort index: 0
Inputs types: [TensorType(float64, (None, None)), TensorType(float64, ()), TensorType(float64, (None, None)), TensorType(float64, (None, None)), TensorType(float64, ())]
Inputs shapes: [(1, 100), (), (10, 6), (6, 100), ()]
Inputs strides: [(800, 8), (), (48, 8), (800, 8), ()]
Inputs values: ['not shown', array(1.), 'not shown', 'not shown', array(1.)]
Outputs clients: [['output']]

HINT: Re-running with most Aesara optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the Aesara flag 'optimizer=fast_compile'. If that does not work, Aesara optimizations can be disabled with 'optimizer=None'.
HINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.
```

### PR DESCRIPTION
**Pull request by ricardoV94**
The `Gemm` `Op` is only applicable when there is no broadcasting between `Z` and `dot(A, B)` in the expression `Z + dot(A, B)`. This PR broadcasts the matrix inputs when the Op was inserted for a mix of row / column matrices that were only known to be general matrices during compilation.

Closes #984

### CODE REVIEW

**Comment by ricardoV94** [comment]
Converting to draft because of the performance concerns that this fix entails

**Comment by brandonwillard** [comment]
> Perhaps it makes sense to allow this type of constraint at the type level, more in line with the old `broadcastable` flag? Type shapes would then not only be limited to `(None, int)`, but also allow for a special flag `-1` or `"not1"` to indicate this dimension can be anything other than `1`.

The old `TensorType.broadcastable` is still present in exactly the same form as it was.  The only differences might be in how we want to use and interpret it.

I made a remark about this (i.e. the "old") interpretation of `TensorType.broadcastable` recently [here](https://github.com/aesara-devs/aesara/pull/981/files#diff-c8a387500e93e9fbc298ed8a0d0aded0feab10fe294ce3d1aadd4ee301f46bb7R2919-R2923)&mdash;among other places/times throughout our work.  The problem with the "old"/strict interpretation is that it puts extra pressure on `Op.make_node` implementations to both infer and be accurate about the broadcast patterns/static shape information in the `TensorType`s it creates.  We've been dealing with the issues and limitations that arise from this interpretation all throughout this work.

The type constraints you mention are viable, but also really do require a much more clearly defined and implemented type system, and, ultimately, some basic constraint logic.  My push for the broad use of miniKanren is&mdash;in part&mdash;motivated by the availability (and compartmentalization) of such features.

Regardless, why can't we broadcast all the inputs to the GEMM `Op` in the rewrite (or even in the `Op.make_node` or `Op.perform`/`Op.c_code` methods)?

**Comment by ricardoV94** [comment]
> Regardless, why can't we broadcast all the inputs to the GEMM `Op` in the rewrite (or even in the `Op.make_node` or `Op.perform`/`Op.c_code` methods)?

I'll explore that. I didn't plan to mess with blas related Ops, but here we are ^^

In any case I feel that supporting (and enforcing) non size1 type shape might come in handy in a couple of places.

**Comment by ricardoV94** [comment]
Tests are passing

**Review by brandonwillard — DISMISSED** [review]
This looks great.

File: aesara/tensor/blas.py [hunk 1]
```diff
@@ -1022,59 +1033,107 @@ def infer_shape(self, fgraph, node, input_shapes):
...
+            if(PyArray_MoveInto(x_new, %(_x)s) == -1)
             {
-                for (int j = 0; j < Nz[1]; ++j)
-                {
-                    zoutdata[zoi*i + zoj*j] = zdata[zi*i + zj*j];
-                }
+                %(fail)s
             }
+
+            Py_DECREF(%(_x)s);
```
Comment (ricardoV94) [inline aesara/tensor/blas.py:1103]: I don't think I want to do this. Instead I want to get rid of the new broadcasted array when I am done.

File: aesara/tensor/blas.py [hunk 2]
```diff
@@ -1022,59 +1033,107 @@ def infer_shape(self, fgraph, node, input_shapes):
...
+                                "failed to alloc gemm_inplace input");
+                %(fail)s
+            }
+
+            if(PyArray_MoveInto(y_new, %(_y)s) == -1)
+            {
+                %(fail)s
+            }
+
+            Py_DECREF(%(_y)s);
```
Comment (ricardoV94) [inline aesara/tensor/blas.py:1129]: Same here

File: aesara/tensor/blas.py [hunk 3]
```diff
@@ -630,8 +629,10 @@ def c_header_dirs(self, **kwargs):
         {PyErr_SetString(PyExc_NotImplementedError, "type(b) is not double or float"); %(fail)s;}
         """
 
+    # broadcast_xyz = None
+
     check_dims = """
-        if (Nx[0] != Nz[0])
+        if (Nx[0] !=1 && Nz[0] != 1 && Nx[0] != Nz[0])
```
Comment (ricardoV94) [inline aesara/tensor/blas.py:635]: Only alter this for Gemm Op, no idea if other Gemm-related Ops are doing the right thing!

**Comment by ricardoV94** [comment]
I did some sanity checks and I am now more confident that I didn't screw up anything :)

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**